File: python/grasslm/layers.py
"""
Core Grassmann layers for GrassLM.

Implements the Causal Grassmann mixing architecture from https://arxiv.org/abs/2512.19428:
- PluckerEncoder: reduces tokens to low-rank space, computes Plucker coordinates
- GrassmannMixing: multi-scale Plucker encoding with gated fusion
- GrassmannBlock: full block with mixing, layer norm, FFN, and residuals
"""

import torch
import torch.nn as nn


class PluckerEncoder(nn.Module):
    """
    Encodes local token pairs as points on Gr(2, r) via Plucker coordinates.

    For each position t and a given offset Δ, forms the pair (z_t, z_{t+Δ})
    in reduced space R^r, computes the Plucker embedding in R^{C(r,2)},
    normalizes, and projects back to model dimension.
    """

    def __init__(self, d_model: int = 256, d_reduce: int = 32):
        super().__init__()
        self.d_model = d_model
        self.d_reduce = d_reduce
        self.plucker_dim = d_reduce * (d_reduce - 1) // 2  # C(r, 2) = 496

        self.W_red = nn.Linear(d_model, d_reduce)
        self.W_plu = nn.Linear(self.plucker_dim, d_model)

        # Precompute index pairs (i, j) with i < j for Plucker coordinates
        idx = torch.combinations(torch.arange(d_reduce), r=2)  # (plucker_dim, 2)
        self.register_buffer("idx_i", idx[:, 0])  # (plucker_dim,)
        self.register_buffer("idx_j", idx[:, 1])  # (plucker_dim,)

    def forward(self, h: torch.Tensor, window_offset: int) -> torch.Tensor:
        """
        Compute Plucker features for token pairs at a given offset.

        Args:
            h: Hidden states, shape (B, L, d_model)
            window_offset: Causal offset Δ > 0; pairs position t with t-Δ

        Returns:
            g: Grassmann features, shape (B, L, d_model).
               First Δ positions (with no backward neighbor) are zero-padded.
        """
        B, L, _ = h.shape
        delta = window_offset

        # 1. Reduce to low-rank space
        z = self.W_red(h)  # (B, L, r)

        # 2. Form causal pairs: z_t paired with z_{t-delta} (look backward)
        if delta >= L:
            # No valid pairs exist
            return torch.zeros_like(h)

        z_t = z[:, delta:, :]           # (B, L-Δ, r)  — positions Δ..L-1
        z_td = z[:, :L - delta, :]      # (B, L-Δ, r)  — positions 0..L-Δ-1

        # 3. Compute Plucker coordinates: p[i,j] = z_t[i]*z_td[j] - z_t[j]*z_td[i]
        z_t_i = z_t[:, :, self.idx_i]   # (B, L-Δ, plucker_dim)
        z_t_j = z_t[:, :, self.idx_j]   # (B, L-Δ, plucker_dim)
        z_td_i = z_td[:, :, self.idx_i]  # (B, L-Δ, plucker_dim)
        z_td_j = z_td[:, :, self.idx_j]  # (B, L-Δ, plucker_dim)

        p = z_t_i * z_td_j - z_t_j * z_td_i  # (B, L-Δ, plucker_dim)

        # 4. Normalize
        p_hat = p / p.norm(dim=-1, keepdim=True).clamp(min=1e-8)

        # 5. Project back to model dimension
        g_valid = self.W_plu(p_hat)  # (B, L-Δ, d_model)

        # 6. Pad to full sequence length (first Δ positions have no backward neighbor)
        g = torch.zeros(B, L, self.d_model, device=h.device, dtype=h.dtype)
        g[:, delta:, :] = g_valid

        return g


class GrassmannMixing(nn.Module):
    """
    Multi-scale Grassmann mixing with gated fusion.

    For each offset in window_offsets, computes Plucker features via a
    shared PluckerEncoder, averages across valid offsets per position,
    and fuses with the original hidden states via a learned gate.
    """

    def __init__(
        self,
        d_model: int = 256,
        d_reduce: int = 32,
        window_offsets: list[int] | None = None,
    ):
        super().__init__()
        if window_offsets is None:
            window_offsets = [1]
        self.window_offsets = window_offsets
        self.d_model = d_model

        # Shared encoder across offsets (as per paper: shared W_red)
        self.plucker_encoder = PluckerEncoder(d_model, d_reduce)

        # Gating: takes [h; g] and produces per-dimension gate α
        self.W_gate = nn.Linear(2 * d_model, d_model)

    def forward(self, h: torch.Tensor) -> torch.Tensor:
        """
        Apply Grassmann mixing.

        Args:
            h: Hidden states, shape (B, L, d_model)

        Returns:
            h_mix: Mixed hidden states, shape (B, L, d_model)
        """
        B, L, d = h.shape

        # 1. Compute Plucker features for each offset
        g_sum = torch.zeros(B, L, d, device=h.device, dtype=h.dtype)
        count = torch.zeros(B, L, 1, device=h.device, dtype=h.dtype)

        for delta in self.window_offsets:
            g_delta = self.plucker_encoder(h, delta)  # (B, L, d)
            g_sum = g_sum + g_delta

            # Track how many offsets are valid per position (positions delta..L-1)
            if delta < L:
                count[:, delta:, :] += 1.0

        # 2. Average over valid offsets
        g = g_sum / count.clamp(min=1.0)

        # 3. Gated fusion: α = sigmoid(W_gate([h; g]))
        u = torch.cat([h, g], dim=-1)  # (B, L, 2d)
        alpha = torch.sigmoid(self.W_gate(u))  # (B, L, d)

        # h_mix = α * h + (1 - α) * g
        h_mix = alpha * h + (1 - alpha) * g

        return h_mix


class GrassmannBlock(nn.Module):
    """
    A single Grassmann Transformer block.

    Replaces self-attention with Grassmann mixing, followed by
    layer norm, FFN with GELU, residual connections, and dropout.

    Architecture:
        h -> GrassmannMixing -> Dropout -> LayerNorm -> (+FFN -> Dropout) -> LayerNorm -> out
    with a residual connection around the FFN.
    """

    def __init__(
        self,
        d_model: int = 256,
        d_reduce: int = 32,
        d_ff: int = 1024,
        window_offsets: list[int] | None = None,
        dropout: float = 0.1,
    ):
        super().__init__()
        if window_offsets is None:
            window_offsets = [1]

        self.mixing = GrassmannMixing(d_model, d_reduce, window_offsets)
        self.ln1 = nn.LayerNorm(d_model)
        self.ffn = nn.Sequential(
            nn.Linear(d_model, d_ff),
            nn.GELU(),
            nn.Linear(d_ff, d_model),
        )
        self.ln2 = nn.LayerNorm(d_model)
        self.dropout = nn.Dropout(dropout)

    def forward(self, h: torch.Tensor) -> torch.Tensor:
        """
        Forward pass through the Grassmann block.

        Args:
            h: Input hidden states, shape (B, L, d_model)

        Returns:
            h_out: Output hidden states, shape (B, L, d_model)
        """
        # Grassmann mixing + dropout + layer norm
        h_mix = self.mixing(h)
        h_mix = self.dropout(self.ln1(h_mix))

        # FFN with residual + layer norm
        h_out = self.ln2(h_mix + self.dropout(self.ffn(h_mix)))

        return h_out

File: cpp/src/layers.cpp
#include <grasslm/layers.h>

#include <algorithm>
#include <cmath>
#include <vector>

namespace grasslm {

// --- PluckerEncoder ---

void PluckerEncoder::init(const Tensor& W_red_weight, const Tensor& W_red_bias,
                          const Tensor& W_plu_weight, const Tensor& W_plu_bias,
                          int d_model, int d_reduce) {
    W_red_weight_ = W_red_weight;
    W_red_bias_ = W_red_bias;
    W_plu_weight_ = W_plu_weight;
    W_plu_bias_ = W_plu_bias;
    d_model_ = d_model;
    d_reduce_ = d_reduce;
    plucker_dim_ = d_reduce * (d_reduce - 1) / 2;

    // Precompute index pairs (i, j) with i < j
    idx_i_.clear();
    idx_j_.clear();
    for (int i = 0; i < d_reduce; ++i) {
        for (int j = i + 1; j < d_reduce; ++j) {
            idx_i_.push_back(i);
            idx_j_.push_back(j);
        }
    }
}

Tensor PluckerEncoder::forward(const Tensor& h, int offset) const {
    int L = h.size(0);

    // 1. Reduce to low-rank space: z = h @ W_red^T + bias
    Tensor z = h.linear(W_red_weight_, W_red_bias_);  // (L, d_reduce)

    // 2. No valid pairs if offset >= L
    if (offset >= L) {
        Tensor result({L, d_model_});
        result.zeros();
        return result;
    }

    int valid_len = L - offset;

    // 3. Form causal pairs: z_t paired with z_{t-offset} (look backward)
    Tensor z_t = z.slice(0, offset, L);           // (valid_len, d_reduce)  — positions offset..L-1
    Tensor z_td = z.slice(0, 0, valid_len);       // (valid_len, d_reduce)  — positions 0..L-offset-1

    // 4. Compute Plucker coordinates:
    //    p[l][k] = z_t[l][idx_i[k]] * z_td[l][idx_j[k]]
    //            - z_t[l][idx_j[k]] * z_td[l][idx_i[k]]
    Tensor p({valid_len, plucker_dim_});
    for (int l = 0; l < valid_len; ++l) {
        for (int k = 0; k < plucker_dim_; ++k) {
            int ii = idx_i_[k];
            int jj = idx_j_[k];
            p(l, k) = z_t(l, ii) * z_td(l, jj) - z_t(l, jj) * z_td(l, ii);
        }
    }

    // 5. Normalize: p_hat = p / ||p||_2 (per row, clamped)
    Tensor p_norm = p.norm(-1);  // (valid_len, 1)
    Tensor p_norm_clamped = p_norm.clamp(1e-8f, 1e30f);

    Tensor p_hat({valid_len, plucker_dim_});
    for (int l = 0; l < valid_len; ++l) {
        float inv_norm = 1.0f / p_norm_clamped(l, 0);
        for (int k = 0; k < plucker_dim_; ++k) {
            p_hat(l, k) = p(l, k) * inv_norm;
        }
    }

    // 6. Project back to model dimension: g_valid = p_hat @ W_plu^T + bias
    Tensor g_valid = p_hat.linear(W_plu_weight_, W_plu_bias_);  // (valid_len, d_model)

    // 7. Pad to full sequence length (first offset positions have no backward neighbor)
    Tensor g({L, d_model_});
    g.zeros();
    float* g_data = g.data();
    const float* gv_data = g_valid.data();
    for (int l = 0; l < valid_len; ++l) {
        std::copy(gv_data + l * d_model_, gv_data + (l + 1) * d_model_,
                  g_data + (l + offset) * d_model_);
    }

    return g;
}

// --- GrassmannMixing ---

void GrassmannMixing::init(const PluckerEncoder& encoder,
                           const Tensor& W_gate_weight, const Tensor& W_gate_bias,
                           const std::vector<int>& window_offsets, int d_model) {
    plucker_encoder_ = encoder;
    W_gate_weight_ = W_gate_weight;
    W_gate_bias_ = W_gate_bias;
    window_offsets_ = window_offsets;
    d_model_ = d_model;
}

Tensor GrassmannMixing::forward(const Tensor& h) const {
    int L = h.size(0);
    int d = d_model_;

    // 1. Compute Plucker features for each offset and accumulate
    Tensor g_sum({L, d});
    g_sum.zeros();
    std::vector<float> count(L, 0.0f);

    for (int delta : window_offsets_) {
        Tensor g_delta = plucker_encoder_.forward(h, delta);  // (L, d)
        g_sum.axpy(1.0f, g_delta);

        // Positions delta..L-1 are valid (they have a backward neighbor)
        for (int t = delta; t < L; ++t) {
            count[t] += 1.0f;
        }
    }

    // 2. Average over valid offsets
    Tensor g({L, d});
    float* g_data = g.data();
    const float* gs_data = g_sum.data();
    for (int t = 0; t < L; ++t) {
        float c = std::max(count[t], 1.0f);
        for (int j = 0; j < d; ++j) {
            g_data[t * d + j] = gs_data[t * d + j] / c;
        }
    }

    // 3. Gated fusion: alpha = sigmoid(W_gate([h; g]))
    Tensor u = Tensor::cat({h, g}, 1);                     // (L, 2d)
    Tensor alpha = u.linear(W_gate_weight_, W_gate_bias_);  // (L, d)
    alpha = alpha.sigmoid();

    // 4. h_mix = alpha * h + (1 - alpha) * g
    Tensor h_mix({L, d});
    float* hm_data = h_mix.data();
    const float* h_data = h.data();
    const float* a_data = alpha.data();
    g_data = g.data();
    for (int i = 0; i < L * d; ++i) {
        hm_data[i] = a_data[i] * h_data[i] + (1.0f - a_data[i]) * g_data[i];
    }

    return h_mix;
}

// --- GrassmannBlock ---

void GrassmannBlock::init(const GrassmannMixing& mixing,
                          const Tensor& ln1_weight, const Tensor& ln1_bias,
                          const Tensor& ffn_w1_weight, const Tensor& ffn_w1_bias,
                          const Tensor& ffn_w2_weight, const Tensor& ffn_w2_bias,
                          const Tensor& ln2_weight, const Tensor& ln2_bias,
                          int d_model, int d_ff) {
    mixing_ = mixing;
    ln1_weight_ = ln1_weight;
    ln1_bias_ = ln1_bias;
    ffn_w1_weight_ = ffn_w1_weight;
    ffn_w1_bias_ = ffn_w1_bias;
    ffn_w2_weight_ = ffn_w2_weight;
    ffn_w2_bias_ = ffn_w2_bias;
    ln2_weight_ = ln2_weight;
    ln2_bias_ = ln2_bias;
    d_model_ = d_model;
    d_ff_ = d_ff;
}

Tensor GrassmannBlock::forward(const Tensor& h) const {
    // Grassmann mixing (no dropout at inference)
    Tensor h_mix = mixing_.forward(h);

    // Layer norm 1
    h_mix = h_mix.layernorm(ln1_weight_, ln1_bias_);

    // FFN: Linear(d_model, d_ff) -> GELU -> Linear(d_ff, d_model)
    Tensor ffn_hidden = h_mix.linear(ffn_w1_weight_, ffn_w1_bias_);  // (L, d_ff)
    ffn_hidden = ffn_hidden.gelu();
    Tensor ffn_out = ffn_hidden.linear(ffn_w2_weight_, ffn_w2_bias_);  // (L, d_model)

    // Residual connection around FFN + Layer norm 2
    Tensor h_out = (h_mix + ffn_out).layernorm(ln2_weight_, ln2_bias_);

    return h_out;
}

}  // namespace grasslm
